extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java

```diff
@@ -0,0 +1,747 @@
+      return Actions.exception(e);
+    }
+  }
+
+  /**
+   * Retrieves the list of all Druid schema names. At present, Druid does
+   * not impose security on schemas, only tables within schemas.
+   */
+  @GET
+  @Path("/list/schemas/names")
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response listSchemas(
+      @Context final HttpServletRequest req
+  )
+  {
+    // No good resource to use: we really need finer-grain control.
+    catalog.authorizer().authorizeAccess(ResourceType.STATE, "schemas", Action.READ, req);
+    return Response.ok().entity(catalog.schemaRegistry().names()).build();
+  }
+
+  /**
+   * Retrieves the list of all Druid table names for which the user has at
+   * least read access.
+   */
+  @GET
+  @Path("/list/tables/names")
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response listTables(
+      @Context final HttpServletRequest req
+  )
+  {
+    List<TableId> tables = catalog.tables().list();
+    Iterable<TableId> filtered = AuthorizationUtils.filterAuthorizedResources(
+        req,
+        tables,
+        tableId -> {
+          SchemaSpec schema = catalog.resolveSchema(tableId.schema());
+          if (schema == null) {
+            // Should never occur.
+            return null;
+          }
+          return Collections.singletonList(
+              catalog.authorizer().resourceAction(schema, tableId.name(), Action.READ));
+        },
+        catalog.authorizer().mapper());
+    return Response.ok().entity(Lists.newArrayList(filtered)).build();
+  }
+
+  /**
+   * Retrieves the list of table names within the given schema for which the
+   * user has at least read access. This returns the list of table <i>definitions</i>
+   * which will probably differ from the list of actual tables. For example, for
+   * the read-only schemas, there will be no table definitions.
+   *
+   * @param dbSchema The Druid schema to query. The user must have read access.
+   */
+  @GET
+  @Path("/schemas/{dbSchema}/names")
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response listTables(
+      @PathParam("dbSchema") String dbSchema,
+      @Context final HttpServletRequest req
+  )
+  {
+    Pair<Response, SchemaSpec> result = validateSchema(dbSchema);
+    if (result.lhs != null) {
+      return result.lhs;
+    }
+    SchemaSpec schema = result.rhs;
+    List<String> tables = catalog.tables().list(dbSchema);
+    Iterable<String> filtered = AuthorizationUtils.filterAuthorizedResources(
+        req,
+        tables,
+        name ->
+          Collections.singletonList(
+              catalog.authorizer().resourceAction(schema, name, Action.READ)),
+        catalog.authorizer().mapper());
+    return Response.ok().entity(Lists.newArrayList(filtered)).build();
+  }
+
+  /**
+   * Retrieves the list of all Druid table metadata for which the user has at
+   * least read access.
+   */
+  @GET
+  @Path("/schemas/{dbSchema}/tables")
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response listTableDetails(
```

Comment:
Should we have another version of function that allow user to get the table metadata for all tables in any schema they have acces to? Or the schema can be given as an optional query parameter?

Reply:
makes sense